README.md
# pHTML Doctype [<img src="https://phtml.io/logo.svg" alt="pHTML" width="90" height="90" align="right">][phtml]

[![NPM Version][npm-img]][npm-url]
[![Build Status][cli-img]][cli-url]
[![Support Chat][git-img]][git-url]

[pHTML Doctype] lets you automatically add the doctype in HTML.

```html
<html>
  <title>pHTML Doctype</title>
</html>

<!-- becomes -->

<!doctype html><html>
  <title>pHTML Doctype</title>
</html>
```

## Usage

Transform HTML files directly from the command line:

```bash
npx phtml source.html output.html -p @phtml/doctype
```

### Node

Add [pHTML Doctype] to your project:

```bash
npm install @phtml/doctype --save-dev
```

Use [pHTML Doctype] to process your HTML:

```js
const phtmlDoctype = require('@phtml/doctype');

phtmlDoctype.process(YOUR_HTML /*, processOptions, pluginOptions */);
```

Or use it as a [pHTML] plugin:

```js
const phtml = require('phtml');
const phtmlDoctype = require('@phtml/doctype');

phtml([
  phtmlDoctype(/* pluginOptions */)
]).process(YOUR_HTML /*, processOptions */);
```

[pHTML Doctype] runs in all Node environments, with special instructions for:

| [Node](INSTALL.md#node) | [CLI](INSTALL.md#phtml-cli) | [Eleventy](INSTALL.md#eleventy) | [Gulp](INSTALL.md#gulp) | [Grunt](INSTALL.md#grunt) |
| --- | --- | --- | --- | --- |

## Options

### doctype

The `doctype` option specifies the doctype being added to the page.

```js
phtmlDoctype({ doctype: 'html4' });
```

```html
<html>
  <title>pHTML Doctype</title>
</html>

<!-- becomes -->

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
  <title>pHTML Doctype</title>
</html>
```

Options include case-insensitive, space-insentive variations of the following:

- `HTML` or `HTML 5`
- `HTML 4` or `HTML 4.01 Strict`
- `HTML 4.01 Transitional`
- `HTML 4.01 Frameset`
- `XHTML` or `XHTML 1.1`
- `XHTML 1` or `XHTML 1.0 Strict`
- `XHTML 1.0 Transitional`
- `XHTML 1.0 Frameset`

### replace

The `replace` option determines whether the doctype will be replaced, whether
or not it exists.

```js
phtmlDoctype({ replace: true });
```

```html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
  <title>pHTML Doctype</title>
</html>

<!-- becomes -->

<!doctype html>
<html>
  <title>pHTML Doctype</title>
</html>
```

Optionally, the `replace` option can be given a string, which will work as a
fallback to the [doctype](#doctype) option.

### safe

The `safe` option determines whether the doctype should only be included when a
top level `html`, `head`, or `body` tag is present. By default, this setting is
`false` and the doctype is added regardless.

```js
phtmlDoctype({ safe: true });
```

```html
<custom-element></custom-element>

<!-- becomes -->

<custom-element></custom-element>
```

```html
<html>
  <custom-element></custom-element>
</html>

<!-- becomes -->

<!doctype html>
<html>
  <custom-element></custom-element>
</html>
```

[cli-img]: https://img.shields.io/travis/phtmlorg/phtml-doctype.svg
[cli-url]: https://travis-ci.org/phtmlorg/phtml-doctype
[git-img]: https://img.shields.io/badge/support-chat-blue.svg
[git-url]: https://gitter.im/phtmlorg/phtml
[npm-img]: https://img.shields.io/npm/v/@phtml/doctype.svg
[npm-url]: https://www.npmjs.com/package/@phtml/doctype

[pHTML]: https://github.com/phtmlorg/phtml
[pHTML Doctype]: https://github.com/phtmlorg/phtml-doctype


CHANGELOG.md
# Changes to pHTML Doctype

### 4.1.0 (May 16, 2019)

- Updated: `phtml` to 4.1.0 (minor)

### 4.0.0 (May 6, 2019)

- Updated: `phtml` to 4.0.0 (major)
- Updated: Node 8+ compatibility (major)

### 3.0.0 (March 16, 2019)

- Updated: `phtml` to 3.0.0 (major)

### 2.1.0 (March 10, 2019)

- Added: `safe` option to conditionally add doctype only when an html, head, or
  body tag is present.
- Updated: `phtml` to 2.0.3 (patch)

### 2.0.0 (February 18, 2019)

- Updated: `phtml` to 2.0.1 (major)

### 1.0.0 (February 10, 2019)

- Initial version


src/index.js
import phtml from 'phtml';

export default new phtml.Plugin('phtml-doctype', opts => {
	const doctype = doctypes[formatDoctype(Object(opts).doctype)] || doctypes[formatDoctype(Object(opts).replace)] || doctypes['html5'];
	const isReplacingDoctype = 'replace' in Object(opts) ? Boolean(opts.replace) : true;
	const safe = 'safe' in Object(opts) ? Boolean(opts.safe) : false;

	return root => {
		const firstNode = root.first;
		const hasDoctype = firstNode && firstNode.type === 'doctype';
		const newDoctypeNode = isReplacingDoctype || !hasDoctype ? new phtml.Doctype(doctype) : null;
		const hasSafeTag = !safe || root.nodes.some(node => safeTagRegExp.test(node.name));

		if (hasSafeTag) {
			if (firstNode) {
				if (hasDoctype) {
					if (isReplacingDoctype) {
						firstNode.replaceWith(newDoctypeNode);
					}
				} else {
					firstNode.before(newDoctypeNode);
				}
			} else {
				root.append(newDoctypeNode);
			}
		}
	};
});

const safeTagRegExp = /^(body|head|html)$/;

const formatDoctype = doctype => String(doctype || '').trim().toLowerCase().replace(/\s/g, '')

const doctypes = {
	'html': {
		name: 'html'
	},
	'html5': {
		name: 'html'
	},
	'html4': {
		doctype: 'DOCTYPE',
		name: 'HTML',
		publicId: 'PUBLIC "-//W3C//DTD HTML 4.01//EN"',
		systemId: '"http://www.w3.org/TR/html4/strict.dtd"'
	},
	'html4.01strict': {
		doctype: 'DOCTYPE',
		name: 'HTML',
		publicId: 'PUBLIC "-//W3C//DTD HTML 4.01//EN"',
		systemId: '"http://www.w3.org/TR/html4/strict.dtd"'
	},
	'html4.01transitional': {
		doctype: 'DOCTYPE',
		name: 'HTML',
		publicId: 'PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"',
		systemId: '"http://www.w3.org/TR/html4/loose.dtd"'
	},
	'html4.01frameset': {
		doctype: 'DOCTYPE',
		name: 'HTML',
		publicId: 'PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN"',
		systemId: '"http://www.w3.org/TR/html4/frameset.dtd"'
	},
	'xhtml': {
		doctype: 'DOCTYPE',
		name: 'html',
		publicId: 'PUBLIC "-//W3C//DTD XHTML 1.1//EN"',
		systemId: '"http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd"'
	},
	'xhtml1.1': {
		doctype: 'DOCTYPE',
		name: 'html',
		publicId: 'PUBLIC "-//W3C//DTD XHTML 1.1//EN"',
		systemId: '"http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd"'
	},
	'xhtml1': {
		doctype: 'DOCTYPE',
		name: 'html',
		publicId: 'PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"',
		systemId: '"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"'
	},
	'xhtml1.0strict': {
		doctype: 'DOCTYPE',
		name: 'html',
		publicId: 'PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"',
		systemId: '"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"'
	},
	'xhtml1.0transitional': {
		doctype: 'DOCTYPE',
		name: 'html',
		publicId: 'PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"',
		systemId: '"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"'
	},
	'xhtml1.0frameset': {
		doctype: 'DOCTYPE',
		name: 'html',
		publicId: 'PUBLIC "-//W3C//DTD XHTML 1.0 Frameset//EN"',
		systemId: '"http://www.w3.org/TR/xhtml1/DTD/xhtml1-frameset.dtd"'
	}
};
